src/redb_blockstore.rs

```diff
@@ -0,0 +1,227 @@
+use std::path::Path;
+use std::sync::Arc;
+
+use cid::CidGeneric;
+use redb::{
+    CommitError, Database, ReadTransaction, ReadableTable, StorageError, TableDefinition,
+    TableError, TransactionError, WriteTransaction,
+};
+use tokio::task::spawn_blocking;
+
+use crate::{Blockstore, BlockstoreError, Result};
+
+const BLOCKS_TABLE: TableDefinition<'static, &[u8], &[u8]> =
+    TableDefinition::new("BLOCKSTORE.BLOCKS");
+
+/// A [`Blockstore`] implementation backed by a [`redb`] database.
+#[derive(Debug)]
+pub struct RedbBlockstore {
+    db: Arc<Database>,
+}
+
+impl RedbBlockstore {
+    /// Open a persistent [`redb`] store.
+    pub async fn open(path: impl AsRef<Path>) -> Result<Self> {
+        let path = path.as_ref().to_owned();
+
+        let db = spawn_blocking(|| Database::create(path))
+            .await?
+            .map_err(|e| BlockstoreError::FatalDatabaseError(e.to_string()))?;
+
+        Ok(RedbBlockstore::new(Arc::new(db)))
+    }
+
+    /// Open an in memory [`redb`] store.
+    pub fn in_memory() -> Result<Self> {
+        let db = Database::builder()
+            .create_with_backend(redb::backends::InMemoryBackend::new())
+            .map_err(|e| BlockstoreError::FatalDatabaseError(e.to_string()))?;
+
+        Ok(RedbBlockstore::new(Arc::new(db)))
+    }
+
+    /// Create new `RedbBlockstore` with an already opened [`redb::Database`].
+    ///
+    /// # Example
+    /// ```
+    /// # async fn example() -> Result<(), Box<dyn std::error::Error>> {
+    /// # use std::sync::Arc;
+    /// use blockstore::RedbBlockstore;
+    /// use tokio::task::spawn_blocking;
+    ///
+    /// let db = spawn_blocking(|| redb::Database::create("path/to/db")).await??;
+    /// let db = Arc::new(db);
+    /// let blockstore = RedbBlockstore::new(db);
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn new(db: Arc<Database>) -> Self {
+        RedbBlockstore { db }
+    }
+
+    /// Returns the raw [`redb::Database`].
+    ///
+    /// This is useful if you want to pass the database handle to any other
+    /// stores (e.g. [`blockstore`]).
+    pub fn raw_db(&self) -> Arc<Database> {
+        self.db.clone()
+    }
+
+    /// Execute a read transaction.
+    async fn read_tx<F, T>(&self, f: F) -> Result<T>
+    where
+        F: FnOnce(&mut ReadTransaction) -> Result<T> + Send + 'static,
+        T: Send + 'static,
+    {
+        let db = self.db.clone();
+
+        spawn_blocking(move || {
+            let mut tx = db.begin_read()?;
+            f(&mut tx)
+        })
+        .await?
+    }
+
+    /// Execute a write transaction.
+    ///
+    /// If closure returns an error the transaction is aborted, otherwise commited.
```

Comment:
```suggestion
    /// If closure returns an error the store state is not changed, otherwise transaction is commited.
```
Aborting the transaction already suggests roll-back to the previous state, but would't hurt to spell it out